hwilib/devices/bitbox02.py

```diff
@@ -252,9 +258,27 @@ def func(*args, **kwargs):  # type: ignore
             raise exc
         except FirmwareVersionOutdatedException as exc:
             raise DeviceNotReadyError(str(exc))
+        except ValueError as e:
+            raise BadArgumentError(str(e))
 
     return cast(T, func)
 
+class BitBox02Simulator():
+    def __init__(self) -> None:
+        self.client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        ip, port = SIMULATOR_PATH.split(":")
+        self.client_socket.connect((ip, int(port)))
```

Comment:
It throw`ConnectionRefusedError` if it wasn't able to connect to the BitBox02 simulator, so we won't be able to use any other simulator.

Reply:
As far as I saw, Trezor or Keepkey also call `connect` on their simulators without handling any exceptions or `connect_ex`. Do you have a suggestion how to handle this exactly?

Reply:
> As far as I saw, Trezor or Keepkey also call connect on their simulators without handling any exceptions or connect_ex. 

See:
```py
def enumerate(password: Optional[str] = None, expert: bool = False, chain: Chain = Chain.MAIN, allow_emulators: bool = False) -> List[Dict[str, Any]]:
    results = []
    devs = hid.HidTransport.enumerate()
    devs.extend(webusb.WebUsbTransport.enumerate())
    if allow_emulators:
        devs.extend(udp.UdpTransport.enumerate())
    for dev in devs:
```

When calling `enumerate` with `--simulators`, Trezor will call `enumerate` from `UdpTransport`. This function will return nothing if it wasn't able to find a device (simulator) so `devs` is empty.

Maybe you could do something like:
```diff
diff --git a/hwilib/devices/bitbox02.py b/hwilib/devices/bitbox02.py
index 53497c6..7cd6929 100644
--- a/hwilib/devices/bitbox02.py
+++ b/hwilib/devices/bitbox02.py
@@ -186,6 +186,8 @@ def enumerate(password: Optional[str] = None, expert: bool = False, chain: Chain
         devs.append(SIMULATOR_PATH)
     for path in devs:
         client = Bitbox02Client(path=path)
+        if allow_emulators and not client.transport:
+            continue
         if path != SIMULATOR_PATH:
             client.set_noise_config(SilentNoiseConfig())
         d_data: Dict[str, object] = {}
@@ -267,7 +269,11 @@ class BitBox02Simulator():
     def __init__(self) -> None:
         self.client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
         ip, port = SIMULATOR_PATH.split(":")
-        self.client_socket.connect((ip, int(port)))
+        self.connected = True
+        try:
+            self.client_socket.connect((ip, int(port)))
+        except:
+            self.connected = False
 
     def write(self, data: bytes) -> None:
         # Messages from client are always prefixed with HID report ID(0x00), which is not expected by the simulator.
@@ -301,7 +307,10 @@ class Bitbox02Client(HardwareWalletClient):
             self.noise_config = CLINoiseConfig()
         else:
             simulator = BitBox02Simulator()
-            self.transport = u2fhid.U2FHid(simulator)
+            if simulator.connected:
+                self.transport = u2fhid.U2FHid(simulator)
+            else:
+                self.transport = None
         self.device_path = path
 
         # use self.init() to access self.bb02.
```

Reply:
Okay, I did a fix based on this(not the same since this does not pass type checks). Now, in my tests, other devices are also passing the tests.